core/src/upgrade.rs

```diff
@@ -145,8 +148,43 @@ pub trait UpgradeInfo {
     fn protocol_info(&self) -> Self::InfoIter;
 }
 
+pub trait UpgradeProtocols {
+    type Iter: Iterator<Item = Protocol>;
```

Comment:
In what use-case would one want to return an `Iterator` instead of a simple slice?

Reply:
For upgrades such as `Select`, we can't return a slice because the list of protocols is not a continuous block of memory.